fastai/learner.py

```diff
@@ -167,9 +187,17 @@ def fit_gen(self, model, data, layer_opt, n_cycle, cycle_len=None, cycle_mult=1,
 
         if best_save_name is not None:
             callbacks+=[SaveBestModel(self, layer_opt, metrics, best_save_name)]
+            
+        if use_swa:
+            # make a copy of the model to track average weights
+            self.swa_model = copy.deepcopy(model)
```

Comment:
I'm not sure about using `deepcopy`, in terms of either correctness or efficiency. Would `state_dict()` be better? Or grab the params from layer_opt?

Reply:
That's a good question, and I admit I wasn't totally sure if this was the right way to do it. I found a few things from the PyTorch maintainers that made me think that `copy.deepcopy(model)` was the suggested way to copy a PyTorch model:
* [this issue by soumith](https://github.com/pytorch/pytorch/issues/415)
* [this commit by apaszke](https://github.com/pytorch/pytorch/pull/454/commits/9c07f61b021b179e1974f9130cbd02848ef5a345)
* [this forum post](https://discuss.pytorch.org/t/are-there-any-recommended-methods-to-clone-a-model/483)

So then the question is, do we really need a copy of the model? I think it would be possible to perform the algorithm by keeping the average as a list of parameters, rather than an entire model (basically in [this line](https://github.com/fastai/fastai/pull/276/files#diff-6b218dc5c075483a5c9b7a5085e4570dR35) I could have `swa_params` be a list rather than getting them from `swa_model` each time). 

But as far as I've been able to figure it out, whenever we want to do inference, we still need a copy of the model to load that list of parameters into. And if we'd have to load those parameters in every time we want to calculate the validation score, then it seems to make more sense to just have a whole copy of the model from the start.

The authors' implementation similarly has two models around during training (the model being trained and the swa_model keeping track of the averages), but they avoid having to do `deepcopy` because they take in the model constructor function and a dict of kwargs to call it with, so they can just instantiate two models. Because `learn.model` is an already-instantiated model instance, we can't do that in this case.

On the efficiency question, I did `%timeit` for `deepcopy` on `preresnet110` and got: `32.1 ms ± 541 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)`. I don't think that's too bad compared to training time.